synapse/lib/jsonstor.py

```diff
@@ -0,0 +1,380 @@
+
+        pkey = self._pathToPkey(path)
+
+        oldb = self.slab.replace(pkey, buid, db=self.pathdb)
+        if oldb is not None:
+            self._incRefObj(oldb, -1)
+
+        self.slab.put(buid + b'refs', s_msgpack.en(1), db=self.metadb)
+
+        self.dirty[buid] = item
+
+    async def getPathObj(self, path):
+        buid = self._pathToBuid(path)
+        if buid is None:
+            return None
+        return self._getBuidItem(buid)
+
+    def _getBuidItem(self, buid):
+        item = self.dirty.get(buid)
+        if item is not None:
+            return item
+
+        byts = self.slab.get(buid, db=self.itemdb)
+        if byts is not None:
+            return s_msgpack.un(byts)
+
+    def _pathToBuid(self, path):
+        pkey = self._pathToPkey(path)
+        return self.slab.get(pkey, db=self.pathdb)
+
+    async def delPathObj(self, path):
+        '''
+        Remove a path and decref the object it references.
+        '''
+        pkey = self._pathToPkey(path)
+        buid = self.slab.pop(pkey, db=self.pathdb)
+        if buid is not None:
+            self._incRefObj(buid, valu=-1)
+
+    async def setPathLink(self, srcpath, dstpath):
+        '''
+        Add a link from the given srcpath to the dstpath.
+        NOTE: This causes the item at dstpath to be incref'd
+        '''
+        srcpkey = self._pathToPkey(srcpath)
+        dstpkey = self._pathToPkey(dstpath)
+
+        if self.slab.get(srcpkey, db=self.pathdb):
+            raise s_exc.PathExists(path=srcpath)
+
+        buid = self.slab.get(dstpkey, db=self.pathdb)
+        if buid is None:
+            raise s_exc.NoSuchPath(path=dstpath)
+
+        self._incRefObj(buid, valu=1)
+        self.slab.put(srcpkey, buid, db=self.pathdb)
+
+    async def getPathObjProp(self, path, prop):
+
+        item = await self.getPathObj(path)
+        if item is None:
+            return None
+
+        for name in self._pathToTupl(prop):
+            item = item[name]
+
+        return item
+
+    def _pathToPkey(self, path):
+        path = self._pathToTupl(path)
+        return ('\x00'.join(path)).encode()
+
+    def _pathToTupl(self, path):
+
+        if isinstance(path, str):
+            path = path.split('/')
+
+        return path
+
+    def _tuplToPath(self, path):
+        return '/'.join(path)
+
+    def _pkeyToPath(self, pkey):
+        return pkey.decode().split('\x00')
+
+    async def getPathList(self, path):
+        path = self._pathToTupl(path)
+
+        plen = len(path)
+        pkey = self._pathToPkey(path)
+
+        for lkey, buid in self.slab.scanByPref(pkey, db=self.pathdb):
+            yield self._tuplToPath(self._pkeyToPath(lkey)[plen:])
+
+    async def getPathObjs(self, path):
+        path = self._pathToTupl(path)
+
+        plen = len(path)
+        pkey = self._pathToPkey(path)
+
+        for lkey, buid in self.slab.scanByPref(pkey, db=self.pathdb):
+            yield self._pkeyToPath(lkey)[plen:], self._getBuidItem(buid)
```

Comment:
This is returning the paths as a tuple, not sure if that is intended? I think _pkeyToPath should probably be called _pkeyToTupl (or _pkeyToList) to make it clearer.